### ISSUE TITLE
Make timeouts and retries work together

### ISSUE THREAD

**Issue by georgysavva** [googleapis/google-cloud-go#1941]
As I understand, this library takes care of automatic retries, for both http and grpc transports. And it's really a great feature. 

For example here, the retry settings for `pubsub: publish` operation (grpc transport):
https://github.com/googleapis/google-cloud-go/blob/6416cc735b86e81cff3f558da87aee244e204472/pubsub/apiv1/publisher_client.go#L85-L101

And for `storage: create bucket` operation (http transport):
https://github.com/googleapis/google-cloud-go/blob/6416cc735b86e81cff3f558da87aee244e204472/storage/bucket.go#L91

Also it stated in the documentation that I should set timeouts via context and it's also very convenient. Timeouts are crucial to set since all calls in this library result in a network request that can block/hang for a long time in case of a network error. And we need to restrict this call according to the SLA of the operation.  

But I don't understand how those two mechanisms (timeouts and retries) work together in this library. 
Let's consider the following example: 

We expect that a successful operation `storage: create bucket` should complete within 1 sec. So we set 1 sec as timeout to this call. But currently something is wrong with the network, it means that most likely we won't receive a network error right away, since tcp is async protocol. We will hang on the network request, but we lucky, because we have 1 sec timeout and won't hang for a long time. But since we can only set timeout for the whole call to the library, this timeout will restrict the total time we can spend on retries. And if the first attempt was blocked on a network operation and wasted the whole timeout library won't be able to do any retries. 
To recap: if network operation hangs we won't be able to retry it since we have a single timeout for both individual attempt and all retries.

I think we should introduce a timeout for individual attempt a.k.a per try timeout. That will restrict the API call according to the SLA. And the second timeout for all retries to restrict the amount of time we can spend trying to retry, it also can be done via number of retries.

**Comment by georgysavva** [comment]
Hi. Any update on this?

**Comment by tritone** [comment]
Hi @georgysavva , thanks for filing the issue and apologies for the slow response.

You're correct that we perform retries for most operations in the storage client and that context cancellation should be used to limit the length of time for which a retry will be attempted.

I think that it's unlikely that we would implement a feature like what you propose in the near term, at least for the storage client. I haven't come across this specific request before, and I don't think that it is a feature that's available in other languages' clients or in gsutil. it seems like it would be quite complicated to manage two sets of timeouts for all the various operations that use retries, especially for operations like uploads which actually perform multiple GCS requests under the hood. Furthermore, it's not clear to me that killing and quickly retrying requests will improve performance overall if the root issue is that you're experiencing network reliability problems.

As a workaround, you are always able to set a short cancellation timeout on the library call and then wrap it with your own retry using gax. See example here: https://godoc.org/github.com/googleapis/gax-go/v2#example-Backoff

In the longer term I know we have some work in the pipeline to make retry behavior for the storage client more flexible (e.g. allowing a user to customize which errors are retried). However this is still in the design phase, and I'm not sure ultimately whether it will cover what you're asking for here.

@codyoss @hongalex @tbpg , feel free to add more thoughts if you have them.

**Comment by georgysavva** [comment]
Okay. Thanks for confirming that and the advise.
